common/clock/ratelimiter.go

```diff
@@ -0,0 +1,326 @@
+			// the reservation will still be successfully rolled back.
+
+			// ensure we are unlocked, which will not have happened if an err
+			// was returned immediately.
+			once.Do(unlock)
+			// (re)-acquire the latest now value
+			now, unlock := r.lockCorrectNow(now)
+			defer unlock()
+			res.CancelAt(now)
+		}
+	}()
+
+	if !res.OK() {
+		// impossible to wait for some reason, figure out why
+		if err := ctx.Err(); err != nil {
+			// context already errored or raced, just return it
+			return err
+		}
+		// !OK and not already canceled means either:
+		//   1: insufficient burst, can never satisfy
+		//   2: sufficient burst, but longer wait than deadline allows
+		//
+		// 1 must imply 0 burst and finite limit, as we only reserve 1.
+		// 2 implies too many reservations for a finite wait.
+		//
+		// burst alone is sufficient to tell these apart, and a race that changes
+		// burst during this call will at worst lead to a minor miscategorization.
+		burst := r.limiter.Burst()
+		if burst > 0 {
+			return errWaitLongerThanDeadline
+		}
+		return errWaitZeroBurst
+	}
+	delay := res.DelayFrom(now)
+	if delay == 0 {
+		return nil // available token, allow it
+	}
+	if deadline, ok := ctx.Deadline(); ok && now.Add(delay).After(deadline) {
+		return errWaitLongerThanDeadline // don't wait for a known failure
+	}
+
+	once.Do(unlock) // unlock before waiting
+
+	// wait for cancellation or the waiter's turn
+	timer := r.timesource.NewTimer(delay)
+	defer timer.Stop()
+	select {
+	case <-ctx.Done():
+		return ctx.Err()
+	case <-timer.Chan():
+		return nil
+	}
+}
+
+func (r *reservation) Used(used bool) {
+	if !used {
+		now, unlock := r.limiter.lockCorrectNow(r.reservedAt)
+		// lock must be held while canceling, because it updates the limiter's time
+		defer unlock()
+
+		// if no calls interleaved, this will be the same as the reservation time,
+		// and the cancellation will be able to roll back an immediately-available call.
+		//
+		// otherwise, the ratelimiter's time has been advanced, and this may fail to
+		// restore any tokens.
+		r.res.CancelAt(now)
+	}
+	// if used, do nothing.
+	//
+	// this method largely exists so it can be mocked and asserted,
+	// or customized by an implementation that tracks allowed/rejected metrics.
+}
+
+func (r *reservation) Allow() bool {
+	// because this uses a snapshotted DelayFrom time rather than whatever the
+	// "current" time might be, this result will not change as time passes.
+	//
+	// that's intentional because it's more stable... and a bit surprisingly
+	// it's also important, as interleaving calls may advance the limiter's
+	// "now" time, and turn this "would not allow" into a "would allow".
+	//
+	// in high-CPU benchmarks, if you use the limiter's "now", you can sometimes
+	// see this allowing many times more requests than it should, due to the
+	// interleaving time advancement.
+	return r.res.OK() && r.res.DelayFrom(r.reservedAt) == 0
```

Comment:
Found a flaw while benchmarking!

Previously this was using `r.limiter.lockNow()`'s value, and under high concurrency it could allow millions rather than the intended couple-thousand (afaict due to the limiter's internal-"now" advancing to a time where the reservation would be allowed).

Now it's fixed, and I have not seen it misbehave again.